### PR TITLE
Add functionality to generate auxpow blocks in regtest

### PR DESCRIPTION
**Pull request by exp0nge**
This roughly reverts the work done here that disabled auxpow generation in regtest: https://github.com/dogecoin/dogecoin/pull/1431/files\#diff-ccc24453c13307f815879738d3bf00eec351417537fbf10dde1468180cacd2f1R127-R137

This is a pretty severe functionality change since auxpow is critical to Dogecoin and wallet integrators need to be able to parse the extra data in auxpow blocks.
For future wallet integrators: Dogecoin follows similar schemes as Namecoin for the merged mining support and the spec is here: https://en.bitcoin.it/wiki/Merged_mining_specification

---

_Mining_: Enables auxpow block header in regtest
_RPC_: Now supports an optional parameter to generate blocks with the auxpow header

Hi all, I think this would be a good addition back to make wallet integration easier. Let me know what I can improve here to get this in.

### CODE REVIEW

**Comment by patricklodder** [comment]
closing & reopening to trigger CI against 1.14.4-dev

**Comment by patricklodder** [comment]
Good news: there are no regressions, I mined some testnet3 blocks with this without the auxpow flag. Bad news, I have not yet been able to mine with auxpow flag on, and I'm seeing a lot of this in my debug.log:

```
2021-05-04 19:40:21 ERROR: CheckAuxPowProofOfWork : AUX proof of work failed
2021-05-04 19:40:21 ERROR: ProcessNewBlock: AcceptBlock FAILED
```

I'll debug this a bit and see what else is needed - will try to look into this between ongoing fee planning work.

**Comment by exp0nge** [comment]
@patricklodder Did you set the retries to 5000?

**Comment by exp0nge** [comment]
@patricklodder Did you set the retries to 5000?

**Comment by exp0nge** [comment]
Here's a auxpow block that I mined locally (let me know if it seems off):
```
040162009b992ca8c060f77bd60bddac3ee22b86148d7454aad8d7a7fe890d09f58469f3259e4445de4118d5bc62e7605654ad6ee939fd19d007e3812d7c46409ab92562a87a8f60ffff7f200000000001000000010000000000000000000000000000000000000000000000000000000000000000ffffffff2928d886916449c224718a4609742ab9110debf731628cb7a2453eedab928333fcb80100000000000000ffffffff0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000b48bc97e1228dcedde46d2e915b730d910e55c9206b706fa965492afed2bd11d0000000000000000000000000101000000010000000000000000000000000000000000000000000000000000000000000000ffffffff050256080101ffffffff020010a5d4e80000001976a914142b5e7a7f8728324db776a7257200e2d02db1d588ac0000000000000000266a24aa21a9ede2f61c3f71d1defd3fa999dfa36953755c690689799962b48bebd836974e8cf900000000
```

**Comment by patricklodder** [comment]
I use `generate 1 100000000 1` on testnet (not regtest) to see what happens. No luck after roughly 200 attempts yet.

Thanks for the block! Having a look!

**Comment by exp0nge** [comment]
I used regtest, so there's probably some actual problems with doing this in testnet

**Comment by exp0nge** [comment]
@patricklodder any updates regarding this?

File: src/rpc/mining.cpp [hunk 1]
```diff
@@ -124,22 +124,40 @@ UniValue generateBlocks(boost::shared_ptr<CReserveScript> coinbaseScript, int nG
...
-            --nMaxTries;
+        if (!nMineAuxPow) {
+            while (nMaxTries > 0 && pblock->nNonce < nInnerLoopCount && !CheckProofOfWork(pblock->GetPoWHash(), pblock->nBits, Params().GetConsensus(nHeight))) {
+                ++pblock->nNonce;
+                --nMaxTries;
+            }
+        } else {
+            CAuxPow::initAuxPow(*pblock);
+            CPureBlockHeader& miningHeader = pblock->auxpow->parentBlock;
+            while (nMaxTries > 0 && miningHeader.nNonce < nInnerLoopCount && !CheckProofOfWork(miningHeader.GetHash(), pblock->nBits, Params().GetConsensus(nHeight))) {
```
Comment (patricklodder) [inline src/rpc/mining.cpp:135]: `miningHeader.GetHash()` should be `miningHeader.GetPoWHash()` and then it works for me on testnet too!

https://chain.so/block/DOGETEST/1ab471a198f54310c98e5820a4766894f1cec53e654cb03e25082b387918b78b

Comment (exp0nge) [reply src/rpc/mining.cpp]: Sweet, regtest as well:

```0401620015f0efb3eb3f094999beed2ebbd7253d184100e45894b2a0a345b3f30b8ad89dacda17c77120627d4969798138e7cafc1f2602ee75741e707f88f4b68e8f52f248d8ab60ffff7f200000000001000000010000000000000000000000000000000000000000000000000000000000000000ffffffff292811ff2956195106bbde32cc1ffe044c4b25e199e3d585ffc943b43d3950e22c9c0100000000000000ffffffff0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000c1ce4ca223f2e0f951478af443777be87efed1f4b21b31d3c1217c697086ab8a0000000000000000000000000101000000010000000000000000000000000000000000000000000000000000000000000000ffffffff0401380101ffffffff0200203d88792d00001976a914142b5e7a7f8728324db776a7257200e2d02db1d588ac0000000000000000266a24aa21a9ede2f61c3f71d1defd3fa999dfa36953755c690689799962b48bebd836974e8cf900000000```

**Review by patricklodder — APPROVED** [review]
I'm ACK on this. Tested on both linux w/ regtest and OS X with QT on testnet.

**Comment by patricklodder** [comment]
@rnicoll as the original author of the disabling, could you please review? thank you!

**Comment by rnicoll** [comment]
I've got a hunch this is something I copied from earlier code. Code looks good to me, doing a full reindex just to be sure nothing separate broke, then will approve.

**Review by rnicoll — DISMISSED** [review]
utACK

(Technically partly tested, but only for regressions)

**Comment by patricklodder** [comment]
@exp0nge would you mind doing a rebase, squash & sign?

**Review by patricklodder — APPROVED** [review]
👍

**Pull request merged**